### ISSUE TITLE
Android: Point lights don't work with Vulkan + Qualcomm

### ISSUE THREAD

**Issue by bejado** [google/filament#3602]
Point lights don't seem to work in `sample-gltf-viewer` when using the Vulkan backend.

**To Reproduce**

I'm using the QR code branch to make it easy to upload a .glb test file: https://github.com/google/filament/pull/3599

1. Build Filament for Android:

```
./build.sh -c -i -p android release
```

2. Patch `sample-gltf-viewer` to enable the Vulkan backend and turn down IBL intensity:

```
--- a/android/samples/sample-gltf-viewer/src/main/java/com/google/android/filament/gltf/MainActivity.kt
+++ b/android/samples/sample-gltf-viewer/src/main/java/com/google/android/filament/gltf/MainActivity.kt
@@ -30,6 +30,7 @@ import androidx.camera.core.Preview
 import androidx.camera.lifecycle.ProcessCameraProvider
 import androidx.core.content.ContextCompat
 import androidx.lifecycle.LifecycleOwner
+import com.google.android.filament.Engine
 import com.google.android.filament.utils.KtxLoader
 import com.google.android.filament.utils.ModelViewer
 import com.google.android.filament.utils.Utils
@@ -89,7 +90,7 @@ class MainActivity : AppCompatActivity() {

         doubleTapDetector = GestureDetector(applicationContext, doubleTapListener)

-        modelViewer = ModelViewer(surfaceView)
+        modelViewer = ModelViewer(surfaceView, Engine.create(Engine.Backend.VULKAN))

         surfaceView.setOnTouchListener { _, event ->
             modelViewer.onTouchEvent(event)
@@ -256,7 +257,7 @@ class MainActivity : AppCompatActivity() {
         val ibl = "default_env"
         readCompressedAsset("envs/$ibl/${ibl}_ibl.ktx").let {
             scene.indirectLight = KtxLoader.createIndirectLight(engine, it)
-            scene.indirectLight!!.intensity = 30_000.0f
+            scene.indirectLight!!.intensity = 3000.0f
         }
         readCompressedAsset("envs/$ibl/${ibl}_skybox.ktx").let {
             scene.skybox = KtxLoader.createSkybox(engine, it)
```

3. Install on device

```
./gradlew samples:sample-gltf-viewer:InstallDebug
```

4. Run and upload [this glb test file](https://github.com/google/filament/files/6071072/rgb-lights.glb.zip), which contains 3 point lights.

5. Notice the difference between Vulkan / OpenGL:

Vulkan:

![Screenshot_20210301-164027](https://user-images.githubusercontent.com/5298046/109700522-b6d0da80-7b46-11eb-8964-3504586c41c0.png)

OpenGL:

![Screenshot_20210301-164122](https://user-images.githubusercontent.com/5298046/109700538-bcc6bb80-7b46-11eb-9783-b665f816dff5.png)

It's peculiar on Vulkan, there are black smears where the point lights should be:

![Screenshot_20210301-164027](https://user-images.githubusercontent.com/5298046/109700787-17601780-7b47-11eb-85b9-4610afbae0f2.png)

**Smartphone (please complete the following information):**
 - Device: Pixel 4a
 - OS: Android 11 userdebug RQ1A.210205.004

**Comment by prideout** [comment]
Thanks!

**Comment by bejado** [comment]
This feels like a `NaN` bug. I'll try capturing a frame in AGI and see if that reveals anything.

**Comment by prideout** [comment]
I also see the following validation error, looks like we're setting the 7th miplevel in a texture that has only 6 mips?

vkCreateImageView: pCreateInfo->subresourceRange.baseMipLevel + .levelCount (= 0 + 7 = 7) is greater than the mip level count of the image (i.e. greater than 6). The Vulkan spec states: If subresourceRange.levelCount is not VK_REMAINING_MIP_LEVELS, subresourceRange.baseMipLevel + subresourceRange.levelCount must be less than or equal to the mipLevels specified in VkImageCreateInfo when image was created

**Comment by romainguy** [comment]
What's the texture in this case? Bloom? IBL?

**Comment by prideout** [comment]
It's StructureMipmap, just an off by one error when calling setMinMaxLevels, I'm making a PR for it.

Sorry for the diversion, it's probably not related to the actual issue that Ben is reporting.

**Comment by romainguy** [comment]
Well it's used for SSAO so it could cause bad things related to lighting

**Comment by pixelflinger** [comment]
bug still happening as of today

**Comment by prideout** [comment]
This is "fixed" by adding `-g` to matc.  We tried updating spirv-tools and glslang, but these updates did not fix the issue.

**Comment by prideout** [comment]
Can be worked around by commenting out the following:

    .RegisterPass(CreateLocalAccessChainConvertPass())
    
  This however is quite detrimental to optimization, so I'll try to find the actual GLSL segment.

**Comment by prideout** [comment]
After staring at the SPIRV for the good and bad cases, I realized they were equivalent.

Then I tried on a mali and voila, the problem does not exist.

So, this isn't an optimizer issue at all, it's a Qualcomm issue.  (which reminds me a bit of https://github.com/google/filament/issues/3027, wherein the optimizer triggers a qcom bug)

It would be nice to find a workaround, which is why I haven't closed this yet...
